jerry-core/vm/vm.c

```diff
@@ -2010,10 +2010,49 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
         {
           if (ecma_are_values_integer_numbers (left_value, right_value))
           {
-            ecma_integer_value_t left_integer = (ecma_integer_value_t) left_value;
-            ecma_integer_value_t right_integer = (ecma_integer_value_t) right_value;
+            bool is_less = (ecma_integer_value_t) left_value < (ecma_integer_value_t) right_value;
+
+            if (*byte_code_p <= CBC_BRANCH_IF_TRUE_BACKWARD_3 && *byte_code_p >= CBC_BRANCH_IF_TRUE_BACKWARD)
+            {
+              byte_code_start_p = byte_code_p++;
+              branch_offset = 0;
+
+              switch (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p))
+              {
+                case 1:
+                {
+                  branch_offset = *(byte_code_p++);
+                  break;
+                }
+                case 3:
+                {
+                  branch_offset = *(byte_code_p++);
+                  /* FALLTHRU */
+                }
+                default:
+                {
+                  JERRY_ASSERT (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) == 2
+                                || CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) == 3);
+
+                  branch_offset <<= 8;
+                  branch_offset |= *(byte_code_p++);
+                  branch_offset <<= 8;
+                  branch_offset |= *(byte_code_p++);
+                  break;
+                }
+              }
```

Comment:
OK, let's try another question. So, this is basically a loop unrolling. Have you tried it this way? (Seems a bit more logical to me.)

```c
JERRY_ASSERT (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) >= 1
              || CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) <= 3);
switch (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p))
{
  case 3:
  {
    branch_offset = *(byte_code_p++);
    branch_offset <<= 8;
    /* FALLTHRU */
  }
  case 2:
  {
    branch_offset |= *(byte_code_p++);
    branch_offset <<= 8;
    /* FALLTHRU */
  }
  default:
  {
    branch_offset |= *(byte_code_p++);
    break;
  }
}
```

Reply:
I haven't tried it this way. I followed this part of the code when I wrote my patch:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/vm/vm.c#L852
But yes, it seems more logical. I will test it. 
Should I change the referenced code too?

Reply:
I think that @zherczeg should rather comment on this.

Reply:
The 1 byte argument has the highest chance. If you want to change it try something like:

```
branch_offset = *(byte_code_p++);
if (unlikely (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) > 1)) {
   handle case 2 and 3.
}
```

Reply:
This could be the most compact (and potentially also the fastest) form then:

```c
branch_offset_length = CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p);
JERRY_ASSERT (branch_offset_length >= 1 && branch_offset_length <= 3);
branch_offset = *(byte_code_p++);
if (unlikely (branch_offset_length != 1))
{
  branch_offset <<= 8;
  branch_offset |= *(byte_code_p++);
  if (unlikely (branch_offset_length == 3))
  {
    branch_offset <<= 8;
    branch_offset |= *(byte_code_p++);
  }
}
```
@zherczeg opinion?